roadmap/implementers-guide/src/node/disputes/dispute-coordinator.md

```diff
@@ -4,10 +4,324 @@ The coordinator is the central subsystem of the node-side components which parti
+coordinator will be considered for slashing. While the backing group always gets
+slashed, a serious attack attempt will likely also consist of malicious approval
+checkers which will cast approval votes, although the candidate is invalid. If
+we did not import those votes, those nodes would likely cast an `invalid` explicit
+vote as part of the dispute in addition to their approval vote and thus avoid a
+slash. With the 2/3rd honest assumption it seems unrealistic that malicious
+actors will keep sending approval votes once they became aware of a raised
+dispute. Hence the most crucial approval votes to import are the early ones
+(tranche 0), to take into account network latencies and such we still want to
+import approval votes at a later point in time as well (in particular we need to
+make sure the dispute can conclude, but more on that later).
+
+As mentioned already previously, importing votes is most efficient when batched.
+At the same time approval voting and disputes are running concurrently so
+approval votes are expected to trickle in still, when a dispute is already
+ongoing.
+
+Hence, we have the following requirements for importing approval votes:
+
+1. Only import them when there is a dispute, because otherwise we are
+   wasting lots of resources _always_ for the exceptional case of a dispute.
+2. Import votes batched when possible, to avoid quadratic import complexity.
+3. Take into account that approval voting is still ongoing, while a dispute is
+   already running.
+
+With a design where approval voting sends votes to the dispute-coordinator by
+itself, we would need to make approval voting aware of ongoing disputes and
+once it is aware it could start sending all already existing votes batched and
+trickling in votes as they come. The problem with this is, that it adds some
+unnecessary complexity to approval voting and also we might still import most of
+the votes unbatched, but one-by-one, depending on what point in time the dispute
+was raised.
+
+Instead of the dispute coordinator telling approval-voting that a dispute is
+ongoing for approval-voting to start sending votes to the dispute coordinator,
+it would make more sense if the dispute-coordinator would just ask
+approval-voting for votes of candidates that are currently disputed. This way
+the dispute-coordinator can also pick the time when to ask and we can therefore
+maximize the amount of batching.
+
+Now the question remains, when should the dispute coordinator ask
+approval-voting for votes? As argued above already, querying approval votes at
+the beginning of the dispute, will likely already take care of most malicious
+votes. Still we would like to have a record of all, if possible. So what are
+other points in time we might query approval votes?
+
+In fact for slashing it is only relevant to have them once the dispute
+concluded, so we can query approval voting the moment the dispute concludes!
+There are two potential caveats with this though:
```

Comment:
I see 2 issues.

1. this introduces a race condition - Approval-Voting only keeps around votes until finality. After a dispute concludes, the chain can be finalized. This is based off of off-chain (i.e. non-consensus) information, in particular, the dispute votes that have been seen by the local node as gathered through the return value of `DisputeCoordinatorMessage::DetermineUndisputedChain` in `relay_chain_selection.rs`. The block could be finalized with only f+1 honest validators having seen the dispute conclude, which means that whatever our scheme for posting bad votes is, we need to ensure that any subset of f+1 honest validators can supply them to the chain.
2. The approval-voting DB doesn't persist the signatures of approval-votes. It just keeps a per-candidate bitfield. This was deliberately done to save space (and later undermined with the dispute-coordinator), but it means that the approval-voting DB doesn't have a way to provide the signed approval messages on a candidate requested after-the-fact.

Reply:
(2) is just an engineering issue but worth considering.

(1) is maybe not that important because I think the Gambler's ruin still holds as long as at least 1 validator gets 100% slashed (this is guaranteed by the nature of the dispute) and deep network-level attacks on the entire network to prevent some messages from being seen by specific validators and having those validators finalize a specific block would be extremely difficult - of course, worthwhile, because anyone pulling off that attack even partially would save themselves tens of millions of dollars per attack attempt. However, if we are relaxing this it's best to be explicit about the adjustments to the security model that are being made.

Reply:
Covered in guide and created a ticket: https://github.com/paritytech/polkadot/issues/5864 , together with the existing [vote recovery ticket](https://github.com/paritytech/polkadot/issues/3398) this should be covered. Thanks for those great insights, they made the implementation more messy than I would have liked unfortunately. While I would be eager to come up with something better, I will in the interest of time and because no real security guarantees get violated (we might only miss on slashing some/but never all offenders in some edge cases),  defer resolution to those tickets for now.